File: swift/Ice/filesystem/Sources/Server/MDirectory.swift
// Copyright (c) ZeroC, Inc.

import Ice

/// Provides an in-memory implementation of the Slice interface Directory.
// This class reuses the implementation of Node provided by MNode.
final class MDirectory: MNode, Directory {
    private var contents: [NodePrx] = []

    // Implements Directory.list.
    func list(current _: Ice.Current) -> [NodePrx?] {
        return contents
    }

    // Adds a non-nill node proxy to this directory.
    func addChild(_ child: NodePrx) {
        contents.append(child)
    }

    // Since we define a `dispatch` method in MNode, we need to override it here. Otherwise, we'd use the base class
    // `dispatch`, not the one provided by generated Directory protocol extension.
    override func dispatch(_ request: sending Ice.IncomingRequest) async throws -> Ice.OutgoingResponse {
        // We implement dispatch by reusing the implementation provided by the generated Directory protocol extension.
        try await Self.dispatch(self, request: request)
    }
}

File: java/Ice/filesystem/server/src/main/java/com/example/ice/filesystem/server/MDirectory.java
// Copyright (c) ZeroC, Inc.

package com.example.ice.filesystem.server;

import com.zeroc.Ice.Current;

import java.util.ArrayList;
import java.util.List;

/**
 * Provides an in-memory implementation of the Slice interface Directory.
 */
public class MDirectory extends Node implements Directory {
    private ArrayList<NodePrx> _contents = new ArrayList<>();

    /**
     * Constructs an {@code MDirectory}.
     * @param name the name of this directory
     */
    MDirectory(String name)
    {
        super(name);
    }

    @Override
    public List<NodePrx> list(Current current)
    {
        return _contents;
    }

    /**
     * Adds a node to this directory.
     * @param child the node proxy to add
     */
    void addChild(NodePrx child)
    {
        _contents.add(child);
    }
}
